### ISSUE TITLE
Codegen not working on 2.0.0-preview3-20170913 for netcoreapp2.0

### ISSUE THREAD

**Issue by ChrisBellew** [dotnet/orleans#3409]
Hi there,

Great work guys and girls. Orleans is such a great platform and I hope it keeps growing!

Just wanted to report I am having problems with preview3 on netcoreapp2.0. My project is failing codegen with the following exception:

```
System.BadImageFormatException: Could not load file or assembly 'System.Runtime, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. Reference assemblies should not be loaded for execution.  They can only be loaded in the Reflection-only loader context. (Exception from HRESULT: 0x80131058)
   at System.ModuleHandle.ResolveType(RuntimeModule module, Int32 typeToken, IntPtr* typeInstArgs, Int32 typeInstCount, IntPtr* methodInstArgs, Int32 methodInstCount, ObjectHandleOnStack type)
   at System.ModuleHandle.ResolveTypeHandleInternal(RuntimeModule module, Int32 typeToken, RuntimeTypeHandle[] typeInstantiationContext, RuntimeTypeHandle[] methodInstantiationContext)
   at System.Reflection.RuntimeModule.ResolveType(Int32 metadataToken, Type[] genericTypeArguments, Type[] genericMethodArguments)
Exc level 2: System.BadImageFormatException: Could not load file or assembly 'file:///C:\Users\cjbel\.nuget\packages\microsoft.netcore.app\2.0.0\ref\netcoreapp2.0\System.Runtime.dll' or one of its dependencies. Reference assemblies should not be loaded for execution.  They can only be loaded in the Reflection-only loader context. (Exception from HRESULT: 0x80131058)
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at Orleans.CodeGeneration.GrainClientGenerator.ReferenceResolver.ResolveAssembly(Object sender, ResolveEventArgs args)
   at System.AppDomain.OnAssemblyResolveEvent(RuntimeAssembly assembly, String assemblyFullName)
Exc level 3: System.BadImageFormatException: Cannot load a reference assembly for execution.
```

I have checked and System.Runtime exists on disk at that location. My file version is 4.6.25519.3, which is different to 4.2.0.0 but I'm not sure if that's the problem or it's a red herring. 

I have the following `.csproj`:

```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp2.0</TargetFramework>
    <RuntimeIdentifiers>ubuntu.14.04-x64</RuntimeIdentifiers>
    <DockerComposeProjectPath>..\docker-compose.dcproj</DockerComposeProjectPath>
  </PropertyGroup>
  
  <PropertyGroup>
    <ServerGarbageCollection>true</ServerGarbageCollection>
  </PropertyGroup>

  <ItemGroup>
    <Folder Include="wwwroot\" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore" Version="1.1.2" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="1.1.2" />
    <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="1.1.2" />
    <PackageReference Include="Microsoft.Orleans.Core" Version="2.0.0-preview3-20170913" />
    <PackageReference Include="Microsoft.Orleans.OrleansCodeGenerator" Version="2.0.0-preview3-20170913" />
    <PackageReference Include="Microsoft.Orleans.OrleansCodeGenerator.Build" Version="2.0.0-preview3-20170913" />
    <PackageReference Include="Microsoft.Orleans.OrleansProviders" Version="2.0.0-preview3-20170913" />
    <PackageReference Include="Microsoft.Orleans.OrleansRuntime" Version="2.0.0-preview3-20170913" />
  </ItemGroup>
  <ItemGroup>
    <DotNetCliToolReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="1.0.1" />
  </ItemGroup>

</Project>
```

If anyone thinks they can help me I would really appreciate it. Thanks in advance!

Chris

**Comment by brhinescot** [comment]
I get the same error on build in a dotnet core 2.0 class library. Here is my .csproj file.

``` xml

<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Orleans.Core" Version="2.0.0-preview3-20170913" />
    <PackageReference Include="Microsoft.Orleans.Core.Abstractions" Version="2.0.0-preview3-20170913" />
    <PackageReference Include="Microsoft.Orleans.OrleansCodeGenerator" Version="2.0.0-preview3-20170913" />
    <PackageReference Include="Microsoft.Orleans.OrleansCodeGenerator.Build" Version="2.0.0-preview3-20170913" />
    <PackageReference Include="Orleans.FiniteStateMachine" Version="1.0.0" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\ReaService.Orleans.Definition\ReaService.Orleans.Definition.csproj" />
  </ItemGroup>

</Project>

```

My System.Runtime.dll also exists and has the same version number, 4.6.25519.3

**Comment by ReubenBond** [comment]
Remove the `Microsoft.Orleans.OrleansCodeGenerator.Build` package reference - unfortunately build-time codegen is not supported on .NET Core on non-windows machines yet. We're working on it, though.

**Comment by ChrisBellew** [comment]
Oh ok @ReubenBond, íll look forward to it then, could you please point we towards an alternative if there is one? (manual codegen?)

**Comment by sergeybykov** [comment]
Try adding `Microsoft.Orleans.OrleansCodeGenerator` (run time codegen) as a workaround.

**Comment by brhinescot** [comment]
Thanks Sergey, that solved the issue. I'm hosting Orleans in Service Fabric and added `Microsoft.Orleans.OrleansCodeGenerator` to the Stateless Service project.

**Comment by sergeybykov** [comment]
Good. In 2.0.0-beta (hopefully, very soon) build time codegen should work on non-Windows systems, and the workaround shouldn't be necessary anymore. We are planning to remove run time codegen soon.

**Comment by sergeybykov** [comment]
Resolved via #3445.
